Issue/#639 Show ban duration to user with higher resolution

46bit opened the pull request:
Hi. This is my first PR to FAF.

This PR addresses #639: "Show ban duration to user with higher resolution." I think it would be better to implement #640 in the long term, but I don't know how to do that yet and this PR still gives an improvement.

The tests outline the new behaviour, and test more of what was already there. A temporarily banned user will now see a message saying:

```
You are banned from FAF for 1 day and 8.5 hours.
Reason :
[ban reason here]
```

I haven't actually been banned myself 😇 but this seemed like a good first contribution.

Closes #639.

Askaholic commented:
Awesome! I totally agree that this is a worthwile improvement, and also that it should be replaced in the future :P. I'll take a closer look at it tonight.

46bit commented:
@Askaholic Brilliant. I had quite a few rounds of fixing CI 🙈 but this is ready for review now.

Askaholic commented on tests/unit_tests/test_lobbyconnection.py:
```diff
@@ -949,3 +950,29 @@ async def test_check_policy_conformity_fatal(lobbyconnection, policy_server):
+    await lobbyconnection.abort_connection_if_banned()
+
+    lobbyconnection.player.id = 203 # test user who is permabanned
+    with pytest.raises(ClientError) as banned_error:
```
Hah, I didn't even know pytest yields the error back to you

46bit commented on server/lobbyconnection.py:
```diff
@@ -391,7 +391,7 @@ async def check_user_login(self, conn, username, password):
                                player_id, username, self.session)
 
-            await self.send_ban_message_and_abort(ban_expiry - now, ban_reason)
+            await send_ban_message_and_abort(ban_expiry - now, ban_reason)
```
Done

Askaholic commented on server/lobbyconnection.py:
```diff
@@ -1079,15 +1079,20 @@ async def abort_connection_if_banned(self):
+            raise_ban_error(ban_expiry - now, data[ban.c.reason])
+
+def raise_ban_error(ban_duration, ban_reason):
+    raise ClientError((f"You are banned from FAF {ban_duration_text(ban_duration)}.\n "
```
Sorry, I thought I commented on this earlier, but I must have forgotten to click the "Add" button or something. The client actually expects these messages to be HTML
```suggestion
    raise ClientError((f"You are banned from FAF {ban_duration_text(ban_duration)}.</br> "
```
right now it looks like this (minus the white background which has something to do with linux):
![image](https://user-images.githubusercontent.com/17941539/91335536-10370600-e7c0-11ea-95b8-c42fd9914d4d.png)
The first one happens to line up correctly, but you can see there is no newline after the reason.

Askaholic replied:
Oh I keep clicking the wrong button I guess. I keep clicking "Add single comment" and not the green button which adds it to my review...

46bit replied:
Hopefully this is fixed now :)

Askaholic commented on tests/unit_tests/test_lobbyconnection.py:
```diff
@@ -949,3 +950,29 @@ async def test_check_policy_conformity_fatal(lobbyconnection, policy_server):
+    lobbyconnection.player.id = 204 # test user who is banned for another 46 hours
+    with pytest.raises(ClientError) as banned_error:
+        await lobbyconnection.abort_connection_if_banned()
+    assert re.match('You are banned from FAF for 1 day and 2[0-3]\.[0-9]+ hours.\n Reason :\n Test ongoing ban with 46 hours left', banned_error.value.message)
```
I think you want a raw string
```suggestion
    assert re.match(r'You are banned from FAF for 1 day and 2[0-3]\.[0-9]+ hours\.\n Reason :\n Test ongoing ban with 46 hours left', banned_error.value.message)
```

Askaholic replied:
Oh, also use double quotes for strings please

46bit replied:
Thanks--I haven't used Python in a long time so a lot of this is new to me.

Askaholic reviewed (commented):
I was getting a warning when running the tests:
```
tests/unit_tests/test_lobbyconnection.py:977: DeprecationWarning: invalid escape sequence \.
    assert re.match('You are banned from FAF for 1 day and 2[0-3]\.[0-9]+ hours.\n Reason :\n Test ongoing ban with 46 hours left', banned_error.value.message)
```

46bit commented:
@Askaholic I think I've addressed everything.

I also refactored a new `BanError` type and started outputting the ban reason and expiry time for future use. Unfortunately the test suite is failing with some sort of problem parsing the protocol :/

Askaholic commented on server/lobbyconnection.py:
```diff
@@ -56,6 +56,31 @@ def __init__(self, message, recoverable=True, *args, **kwargs):
 
+class BanError(Exception):
+    """
+    Represents a BanError
```
I don't think this comment adds any value :P

Askaholic commented on server/lobbyconnection.py:
```diff
@@ -56,6 +56,31 @@ def __init__(self, message, recoverable=True, *args, **kwargs):
+        ban_duration = self.ban_expiry - datetime.now()
+        if ban_duration.days > 365 * 100:
+            return "forever"
+        humanized_ban_duration = humanize.precisedelta(ban_duration, minimum_unit='hours')
```
Right now the code is still pretty inconsistent on quote usage, but I've been trying to convert everything over to double quotes (unless the string contains a double quote).
```suggestion
        humanized_ban_duration = humanize.precisedelta(ban_duration, minimum_unit="hours")
```

46bit replied:
I really think this should be done later as a big PR on the whole codebase rather than trying to get it right here.

46bit replied:
(Am volunteering.)

Askaholic replied:
The tricky thing about doing it all at once is that it could cause conflicts with other PR's. If you want to take a stab at that, go ahead, maybe it won't be too bad. But that still does't justify committing new code with single quotes ;)

Askaholic replied:
It's even coding style bullet number 1 as listed in the contributing document: [CONTRIBUTING.md](https://github.com/FAForever/server/blob/develop/CONTRIBUTING.md)

46bit replied:
cool, pushed again. opened a PR for mass-replacing too to see if it's workable: #648.

Askaholic commented on server/lobbyconnection.py:
```diff
@@ -166,6 +191,15 @@ async def on_message_received(self, message):
+                'command': 'notice',
+                'style': 'error',
+                'text': ex.message(),
+                'ban_reason': ex.ban_reason,
+                'ban_expiry': ex.ban_expiry.isoformat()
```
It's a bit weird attaching these fields to the `notice` message, and the way that the client is written does not lend its self well to having dynamic messages (every message gets deserialized into a class, so the class needs to have members defined for all expected fields). So IMO, it's probably better to hold off on this until we add a dedicated ban message, otherwise if someone tries to implement it, it will turn into a mess.

Askaholic commented on tests/data/test-data.sql:
```diff
@@ -323,7 +324,8 @@ insert into ban (player_id, author_id, reason, level, expires_at, revoke_reason,
   (201, 201, 'Test revoked ban', 'GLOBAL', NULL, now() - interval 1 day),
   (202, 202, 'Test expired ban', 'GLOBAL', now() - interval 1 day, NULL),
-  (203, 203, 'Test permanent ban', 'GLOBAL', now() + interval 1000 year, NULL)
+  (203, 203, 'Test permanent ban', 'GLOBAL', make_timestamptz(3020, 01, 01, 12, 0, 0), NULL),
```
`make_timestampz` doesn't exist I guess. You could just use `TIMESTAMP` instead

Askaholic reviewed (commented):
Yea the refactor is great! The travis build is failing because of the `make_timestampz` function you added to the test data file.

46bit commented:
I've removed exporting the fields since that wasn't helpful. Hopefully good to go now.

Askaholic reviewed (approved):
Perfect!
